### PR TITLE
community: Improve mlflow callback 

### PR DESCRIPTION
**Pull request by serena-ruan**
- **Description:** Allow passing run_id to MLflowCallbackHandler to resume a run instead of creating a new run. Support recording retriever relevant metrics. Refactor the code to fix some bugs.

Please make sure your PR is passing linting and testing before submitting. Run `make format`, `make lint` and `make test` from the root of the package you've modified to check this locally.

See contribution guidelines for more information on how to write/run tests, lint, etc: https://python.langchain.com/docs/contributing/

### CODE REVIEW

**Review by baskaryan — COMMENTED** [review]
cc @akshaya-a

**Comment by serena-ruan** [comment]
cc @harupy @BenWilson2 @dbczumar

**Comment by akshaya-a** [comment]
Thanks serena!

File: libs/community/langchain_community/callbacks/mlflow_callback.py [hunk 1]
```diff
@@ -140,89 +170,84 @@ def __init__(self, **kwargs: Any):
...
         """To log the input pandas dataframe as a html table"""
         self.html(dataframe.to_html(), f"table_{name}")
 
     def html(self, html: str, filename: str) -> None:
         """To log the input html string as html file artifact."""
-        with self.mlflow.start_run(
-            run_id=self.run.info.run_id, experiment_id=self.mlf_expid
-        ):
-            self.mlflow.log_text(html, f"{filename}.html")
+        self.mlflow.log_text(html, f"{filename}.html", run_id=self.run_id)
```
Comment (serena-ruan) [inline libs/community/langchain_community/callbacks/mlflow_callback.py:239]: Those changes depend on changes in MLflow here https://github.com/mlflow/mlflow/pull/10801/files#diff-b33420360cc29a5224c8d3080bbfcaa0ac1c8891416137e13be044fcdfdd26f9 I'll include those changes into 2.10.0 release so after that MLflowCallbackHandler can be used without error

**Comment by serena-ruan** [comment]
@akshaya-a @harupy @efriis Could you please review this PR?

File: libs/community/langchain_community/callbacks/mlflow_callback.py [hunk 2]
```diff
@@ -140,89 +170,83 @@ def __init__(self, **kwargs: Any):
...
-            # User can set other env variables described here
-            # > https://www.mlflow.org/docs/latest/tracking.html#logging-to-a-tracking-server
-
-            experiment_name = get_from_dict_or_env(
-                kwargs, "experiment_name", "MLFLOW_EXPERIMENT_NAME"
-            )
-            self.mlf_exp = self.mlflow.get_experiment_by_name(experiment_name)
-            if self.mlf_exp is not None:
-                self.mlf_expid = self.mlf_exp.experiment_id
+            if kwargs.get("run_id", None) is not None:
```
Comment (harupy) [inline libs/community/langchain_community/callbacks/mlflow_callback.py:173]:
```suggestion
            if "run_id" in kwargs:
```

Comment (harupy) [reply libs/community/langchain_community/callbacks/mlflow_callback.py:173]:
```suggestion
            if run_id := kwargs.get("run_id"):
```

or assignment expression

File: libs/community/langchain_community/callbacks/mlflow_callback.py [hunk 3]
```diff
@@ -140,89 +170,83 @@ def __init__(self, **kwargs: Any):
...
+        """
+        If run_id is provided, it will reuse the run with the given run_id.
+        Otherwise, it starts a new run, auto generates the random suffix for name.
+        """
+        if run_id is None:
+            if name.endswith("-%"):
+                rname = "".join(
+                    random.choices(string.ascii_uppercase + string.digits, k=7)
+                )
+                name = name.replace("%", rname)
```
Comment (harupy) [inline libs/community/langchain_community/callbacks/mlflow_callback.py:207]:
```suggestion
                name = name[:-2] + rname
```

File: libs/community/langchain_community/callbacks/mlflow_callback.py [hunk 4]
```diff
@@ -266,26 +293,21 @@ def __init__(
...
+        try:
+            self.nlp = spacy.load("en_core_web_sm")
+        except OSError:
+            logger.warning(
+                "Run `python -m spacy download en_core_web_sm` "
+                "to download en_core_web_sm"
+            )
+            raise
+
+        self.metrics = {key: 0 for key in mlflow_callback_metrics()}
```
Comment (harupy) [inline libs/community/langchain_community/callbacks/mlflow_callback.py:310]: what does 0 mean here?

File: libs/community/langchain_community/callbacks/mlflow_callback.py [hunk 5]
```diff
@@ -620,14 +705,18 @@ def _create_session_analysis_df(self) -> Any:
         )
         return session_analysis_df
 
+    def _contain_llm_records(self):
+        return len(self.records["on_llm_start_records"]) >= 1
```
Comment (harupy) [inline libs/community/langchain_community/callbacks/mlflow_callback.py:710]:
```suggestion
        return bool(self.records["on_llm_start_records"])
```

File: libs/community/langchain_community/callbacks/mlflow_callback.py (hunk 4 above)
Comment (serena-ruan) [reply libs/community/langchain_community/callbacks/mlflow_callback.py:311]: The original value of all metrics are 0, whenever a certain function is invoked, say on_chain_start, then corresponding metric's value will increase 1 (chain_starts += 1)

File: libs/community/langchain_community/callbacks/mlflow_callback.py [hunk 6]
```diff
@@ -426,7 +458,12 @@ def on_chain_end(self, outputs: Dict[str, Any], **kwargs: Any) -> None:
         chain_ends = self.metrics["chain_ends"]
 
         resp: Dict[str, Any] = {}
-        chain_output = ",".join([f"{k}={v}" for k, v in outputs.items()])
+        if isinstance(outputs, dict):
+            chain_output = ",".join([f"{k}={v}" for k, v in outputs.items()])
+        elif isinstance(outputs, list):
+            chain_output = ",".join([str(output) for output in outputs])
```
Comment (harupy) [inline libs/community/langchain_community/callbacks/mlflow_callback.py:464]:
```suggestion
            chain_output = ",".join(map(str, outputs))
```

File: libs/community/langchain_community/callbacks/mlflow_callback.py [hunk 7]
```diff
@@ -140,89 +170,83 @@ def __init__(self, **kwargs: Any):
...
-            self.mlflow.log_metrics(data)
+        self.mlflow.log_metrics(data, run_id=self.run_id)
 
     def jsonf(self, data: Dict[str, Any], filename: str) -> None:
         """To log the input data as json file artifact."""
-        with self.mlflow.start_run(
-            run_id=self.run.info.run_id, experiment_id=self.mlf_expid
-        ):
-            self.mlflow.log_dict(data, f"{filename}.json")
+        self.mlflow.log_dict(data, f"{self.dir}{filename}.json", run_id=self.run_id)
```
Comment (harupy) [inline libs/community/langchain_community/callbacks/mlflow_callback.py:228]: Can we use `posixpath.path.join` to join paths?

**Review by harupy — APPROVED** [review]
LGTM

**Review by baskaryan — APPROVED** [review]

**Pull request merged**